tests/Neo.UnitTests/Persistence/UT_MemoryClonedCache.cs

```diff
@@ -0,0 +1,127 @@
+//
+// Redistribution and use in source and binary forms with or without
+// modifications are permitted.
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Neo.Persistence;
+using Neo.SmartContract;
+
+namespace Neo.UnitTests.Persistence;
+
+/// <summary>
+/// When adding data to `datacache` <see cref="DataCache"/>,
+/// it gets passed to `snapshotcache` <see cref="SnapshotCache"/> during commit.
+/// If `snapshotcache` <see cref="SnapshotCache"/>commits, the data is then passed
+/// to the underlying store <see cref="IStore"/>.
+/// However, because snapshots <see cref="ISnapshot"/> are immutable, the new data
+/// cannot be retrieved from the snapshot <see cref="ISnapshot"/>.
+///
+/// When deleting data from `datacache` <see cref="DataCache"/>,
+/// it won't exist in `datacache` upon commit, and therefore will be removed from `snapshotcache` <see cref="SnapshotCache"/>.
+/// Upon `snapshotcache` <see cref="SnapshotCache"/>commit, the data is deleted from the store <see cref="IStore"/>.
+/// However, since the snapshot <see cref="ISnapshot"/> remains unchanged, the data still exists in the snapshot.
+/// If you attempt to read this data from `datacache` <see cref="DataCache"/> or `snapshotcache` <see cref="SnapshotCache"/>,
+/// which do not have the data, they will retrieve it from the snapshot instead of the store.
+/// Thus, they can still access data that has been deleted.
+/// </summary>
+[TestClass]
+public class UT_MemoryClonedCache
+{
+    private MemoryStore _memoryStore;
+    private MemorySnapshot _snapshot;
+    private SnapshotCache _snapshotCache;
+    private DataCache _dataCache;
+
+    [TestInitialize]
+    public void Setup()
+    {
+        _memoryStore = new MemoryStore();
+        _snapshot = _memoryStore.GetSnapshot() as MemorySnapshot;
+        _snapshotCache = new SnapshotCache(_snapshot);
+        _dataCache = _snapshotCache.CreateSnapshot();
+    }
+
+    [TestCleanup]
+    public void CleanUp()
+    {
+        _dataCache.Commit();
+        _snapshotCache.Commit();
+        _memoryStore.Reset();
+    }
+
+    [TestMethod]
+    public void SingleSnapshotCacheTest()
+    {
+        var key1 = new KeyBuilder(0, 1);
+        var value1 = new StorageItem([0x03, 0x04]);
+
+        Assert.IsFalse(_dataCache.Contains(key1));
+        _dataCache.Add(key1, value1);
+
+        Assert.IsTrue(_dataCache.Contains(key1));
+        Assert.IsFalse(_snapshotCache.Contains(key1));
+        Assert.IsFalse(_snapshot.Contains(key1.ToArray()));
+        Assert.IsFalse(_memoryStore.Contains(key1.ToArray()));
+
+        // After the data cache is committed, it should be dropped
+        // so its value after the commit is meaningless and should not be used.
+        _dataCache.Commit();
+
+        Assert.IsTrue(_dataCache.Contains(key1));
+        Assert.IsTrue(_snapshotCache.Contains(key1));
+        Assert.IsFalse(_snapshot.Contains(key1.ToArray()));
+        Assert.IsFalse(_memoryStore.Contains(key1.ToArray()));
```

Comment:
I can't agree with this comment (and with related comments in this PR). One of the purposes of DataCache is to serve sequential block's handling in Blockchain (the main part of all Blockchain: OnPersist + transactions handling + PostPersist). And there DataCache usage goes against this comment: we create a single DataCache instance that should be shared between subsequent transaction runs. If transaction is HALTed, then DataCache is being committed and then it's being *reused* for the next transaction processing, so next transactions *can* see the changes made by the previous transaction even after DataCacke commit.

So I think that *reusage after commit* is a substantial part of the DataCache design for *sequential* execution and can't be prohibited like it's said in the comment.

https://github.com/neo-project/neo/blob/c54c2043d8f78c5877bbcb53331b147a3f28440f/src/Neo/Ledger/Blockchain.cs#L442-L461

Parallel execution is different although, and I'm not really sure that it's designed for parallel execution, I'm not that familiar with all DataCache usages.

Reply:
> `I can't agree with this comment `

As i have said everywhere multiple times, the real problem is a committed snapshotcache and clonedcache no longer have the latest store states, that is the reason. You can not agree with what? like any type of execution should be run based on an out dated snapshot? How do you know you can get the correct execution result?